### PR TITLE
Plugin system with MEF

### PR DESCRIPTION
**Pull request by david-driscoll**

### CODE REVIEW

File: src/OmniSharp.Abstractions/Roslyn/OmniSharpWorkspace.cs [hunk 1]
```diff
@@ -13,6 +17,10 @@ public class OmnisharpWorkspace : Workspace
 
         public BufferManager BufferManager { get; private set; }
 
+        public CompositionHost PluginHost { get; private set; }
+
+        public static OmnisharpWorkspace Instance{ get; private set; }
```
Comment (david-driscoll) [inline src/OmniSharp.Abstractions/Roslyn/OmniSharpWorkspace.cs]: I don't like having to have a static property.  Does anyone have any idea how to insert an export into an MEF CompositionHost?

I want to force the value in from the startup class if possible.

File: src/OmniSharp/project.json [hunk 2]
```diff
@@ -4,7 +4,9 @@
         "warningsAsErrors": true
     },
     "dependencies": {
+        "OmniSharp.Abstractions": "1.0.0-*",
         "OmniSharp.Stdio": "1.0.0-*",
+        "OmniSharp.Roslyn.CSharp": "1.0.0-*",
```
Comment (david-driscoll) [inline src/OmniSharp/project.json]: This is just temporary, to force a reference to the assembly, for true plugins we'd have a set of folders to scan for assemblies.

File: src/OmniSharp/Api/v1/Navigation/OmnisharpController.GotoDefinition.cs [hunk 3]
```diff
@@ -16,59 +18,16 @@ public partial class OmnisharpController
         [HttpPost("gotodefinition")]
         public async Task<GotoDefinitionResponse> GotoDefinition(GotoDefinitionRequest request)
         {
-            var quickFixes = new List<QuickFix>();
-
-            var document = _workspace.GetDocument(request.FileName);
-            var response = new GotoDefinitionResponse();
-            if (document != null)
+            var definitions = _workspace.PluginHost.GetExports<IGotoDefintion>();
+            foreach (var instance in definitions)
```
Comment (nosami) [inline src/OmniSharp/Api/v1/Navigation/OmnisharpController.GotoDefinition.cs]: This pattern is going to be the same for all request handlers. Would be nice to abstract it away somehow.

Comment (nosami) [inline src/OmniSharp/Api/v1/Navigation/OmnisharpController.GotoDefinition.cs]: I feel like this whole class could disappear and replaced with RegisterHandler<GotoDefinitionRequest, GotoDefinitionResponse>("gotodefinition") somewhere.

File: src/OmniSharp.Abstractions/Endpoints/IGotoDefinition.cs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+using System.Threading.Tasks;
+using OmniSharp.Models;
+
+namespace OmniSharp.Endpoints
+{
+    public interface IGotoDefintion
+    {
+        Task<bool> IsApplicableTo(GotoDefinitionRequest request);
```
Comment (nosami) [inline src/OmniSharp.Abstractions/Endpoints/IGotoDefinition.cs]: This interface could be generic I think.

Maybe

```
public interface IRequestHandler<TRequest, TResponse>
{
    bool IsApplicableTo(Language language);
   Task<TResponse> Handle(TRequest request);
}
```

File: src/OmniSharp.Roslyn.CSharp/Services/RoslynGotoDefinition.cs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+using OmniSharp.Models;
+
+namespace OmniSharp.Roslyn.CSharp.Services
+{
+    [Export(typeof(IGotoDefintion))]
+    class RoslynGotoDefinition : IGotoDefintion
+    {
+        public Task<bool> IsApplicableTo(GotoDefinitionRequest request)
+        {
+            return Task.FromResult(OmnisharpWorkspace.Instance.GetDocument(request.FileName) != null);
```
Comment (nosami) [inline src/OmniSharp.Roslyn.CSharp/Services/RoslynGotoDefinition.cs]: how about ```return extension.Equals("cs")```

Comment (nosami) [inline src/OmniSharp.Roslyn.CSharp/Services/RoslynGotoDefinition.cs]: The request object seems too heavyweight for this method. Should just be ```Language``` or ```Extension``` for file extension I think. I can't think of a situation where we need more than that.

File: src/OmniSharp.Abstractions/Endpoints/IGotoDefinition.cs (hunk 4 above)
Comment (nosami) [inline src/OmniSharp.Abstractions/Endpoints/IGotoDefinition.cs]: With the right [attributes](https://msdn.microsoft.com/en-us/library/ee155691(v=vs.110).aspx), it might be possible to avoid this method call altogether. Just fetch the gotodefinition handler than can handle C#

File: src/OmniSharp/Api/v1/Navigation/OmnisharpController.GotoDefinition.cs (hunk 3 above)
Comment (david-driscoll) [inline src/OmniSharp/Api/v1/Navigation/OmnisharpController.GotoDefinition.cs]: We could probably just insert everything as custom middleware, I'll look at that, would let us remove most of our endpoints (perhaps not all) from the mvc controller.

File: src/OmniSharp.Roslyn.CSharp/Services/RoslynGotoDefinition.cs (hunk 5 above)
Comment (david-driscoll) [inline src/OmniSharp.Roslyn.CSharp/Services/RoslynGotoDefinition.cs]: I thought about `return extension.Equals("cs")` but then `ScriptCs` and `Cake` crossed my mind, which would also be using the Roslyn Services.

Comment (nosami) [inline src/OmniSharp.Roslyn.CSharp/Services/RoslynGotoDefinition.cs]: `return ["cs", "csx", "cake"].Contains(extension)` :)

File: src/OmniSharp.Abstractions/Endpoints/IGotoDefinition.cs (hunk 4 above)
Comment (david-driscoll) [inline src/OmniSharp.Abstractions/Endpoints/IGotoDefinition.cs]: So we could, but should we?  I suppose it's more `Functional`, lol.

We could do something like...

    public static class OmniSharpEndpoints
    {
        public static Type GotoDefinition {get;} = typeof(Func<GotoDefinitionRequest, Task<GotoDefintionResponse>>)
    }

    class RoslynServices
    {
        [OmniSharpEndpoint(OmniSharpEndpoints.GotoDefinition, Languages.CSharp)]
        public async Task<GotoDefinitionResponse> GotoDefinition(GotoDefinitionRequest request) { ... }
    }

That just feels verbose on the C# side, but may be easier on the F# side of things?  Could we support both?  And additionally have....

    [OmniSharpEndpoint(typeof(IGotoDefinition), Languages.CSharp")]
    class RoslynServices : IGotoDefinition { ... }

Then there is just an extra check to find the correct service, for the given language (which could be thrown behind a cache).

I'm not against the former, it would enforce some extra rules on our part to make sure we
a) never change the delegates
b) make sure we document any changes to the delegates up front.

Note: `OmniSharpEndpointAttribute` would derive from `ExportAttribute` and implement metadata attribute.  On the import side, we would import that metadata with the language.

Comment (david-driscoll) [inline src/OmniSharp.Abstractions/Endpoints/IGotoDefinition.cs]: The one issue I see with the delegate route is that we can't initialize any properties from the MEF host.  So if we need any external services, they will all have to be done on demand in each delegate.

Also the workspace will have to stay static, or become part of the delegate itself, so we can hand it.

Comment (nosami) [inline src/OmniSharp.Abstractions/Endpoints/IGotoDefinition.cs]: Erm, where did the talk of delegates come from? :)

Comment (nosami) [inline src/OmniSharp.Abstractions/Endpoints/IGotoDefinition.cs]: Thinking about this more, the IsApplicableTo logic shouldn't be here at all. It should be at the top level (what types of files does this plugin handle?)

File: src/OmniSharp.Roslyn.CSharp/CSharpLanguage.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+using Microsoft.CodeAnalysis;
+using OmniSharp.Mef;
+
+namespace OmniSharp.Roslyn.CSharp.Services
+{
+    public class LanguagePredicate
+    {
+        private static readonly string[] ValidCSharpExtensions = { "cs", "csx", "cake" };
+        [OmniSharpLanguage(LanguageNames.CSharp)]
+        public Func<string, Task<bool>> IsApplicableTo { get; } = filePath => Task.FromResult(ValidCSharpExtensions.Any(extension => filePath.EndsWith(extension)));
```
Comment (david-driscoll) [inline src/OmniSharp.Roslyn.CSharp/CSharpLanguage.cs]: This was impossible to do as a method, and it's specifically something that isn't supported by the PCL version of MEF.  Delegate Properties on the other hand work just fine.

File: src/OmniSharp/Middleware/Endpoint/EndpointHandler.cs [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        private readonly Type _responseType;
+        private readonly OmnisharpWorkspace _workspace;
+
+        public EndpointHandler(OmnisharpWorkspace workspace, LanguagePredicateHandler languagePredicateHandler, CompositionHost host, EndpointMapItem item)
+        {
+            EndpointName = item.EndpointName;
+            _host = host;
+            _languagePredicateHandler = languagePredicateHandler;
+            _interfaceType = item.InterfaceType;
+            _workspaceDelegateType = WorkspaceFuncType.MakeGenericType(typeof(OmnisharpWorkspace), item.RequestType, TaskType.MakeGenericType(item.ResponseType));
```
Comment (david-driscoll) [inline src/OmniSharp/Middleware/Endpoint/EndpointHandler.cs]: Support for `(OmnisharpWorkspace workspace, TRequest) => TResponse`

File: src/OmniSharp/Middleware/Endpoint/EndpointHandler.cs [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+            _delegateType = FuncType.MakeGenericType(item.RequestType, TaskType.MakeGenericType(item.ResponseType));
```
Comment (david-driscoll) [inline src/OmniSharp/Middleware/Endpoint/EndpointHandler.cs]: Support for `(TRequest) => TResponse`

File: src/OmniSharp/Middleware/Endpoint/EndpointHandler.cs [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+        private readonly Type _requestType;
+        private readonly Type _responseType;
+        private readonly OmnisharpWorkspace _workspace;
+
+        public EndpointHandler(OmnisharpWorkspace workspace, LanguagePredicateHandler languagePredicateHandler, CompositionHost host, EndpointMapItem item)
+        {
+            EndpointName = item.EndpointName;
+            _host = host;
+            _languagePredicateHandler = languagePredicateHandler;
+            _interfaceType = item.InterfaceType;
```
Comment (david-driscoll) [inline src/OmniSharp/Middleware/Endpoint/EndpointHandler.cs]: Support for implementation by interface.

File: src/OmniSharp.Roslyn.CSharp/CSharpLanguage.cs (hunk 6 above)
Comment (nosami) [inline src/OmniSharp.Roslyn.CSharp/CSharpLanguage.cs]: Why the Task?

Comment (david-driscoll) [inline src/OmniSharp.Roslyn.CSharp/CSharpLanguage.cs]: If the plugin is out of process, then we might be waiting for stdio to give us a result, figured this would make that logic easier.  Though thinking about it, maybe it won't be needed.

File: src/OmniSharp.Abstractions/Endpoints.cs [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    {
+        public static EndpointMapItem[] AvailableEndpoints = {
+            EndpointMapItem.Create<GotoDefinition, GotoDefinitionRequest, GotoDefinitionResponse>("/gotodefinition"),
+        };
+
+        public interface GotoDefinition
+        {
+            Task<GotoDefinitionResponse> GotoDefinition(GotoDefinitionRequest request);
+        }
+
```
Comment (nosami) [inline src/OmniSharp.Abstractions/Endpoints.cs]: Could this be

```cs
        public static EndpointMapItem[] AvailableEndpoints = {
            EndpointMapItem.Create<GotoDefinition>("/gotodefinition"),
        };

        public interface Handler<TRequest, TResponse>
        {
            Task<TResponse> Handle(TRequest request);
        }
```

**Comment by david-driscoll** [comment]
keeps getting better!

* Moved the following to their own projects
  * Dnx
  * MsBuild (currently only supporting dnx451)
  * ScriptCs (currently only supporting dnx451)
* Shuffled unit tests with the moves
* Skeleton of how Oop Plugins might work
* Oop Plugins are aligned with the OmniSharp stdio interface, so in theory, OmniSharp [dnxcore] could start OmniSharp [dnx451] to add MSBuild / ScriptCs support.

**Comment by davidfowl** [comment]
Uh interesting 😀

**Comment by david-driscoll** [comment]
Feedback is welcome. :surfer:

**Comment by DustinCampbell** [comment]
This looks really promising. I'm trying to sort out how best to factor our OmniSharp in VS Code to avoid bringing along the "kitchen sink". Having a supported plug-in model would make that easier.

What is your thinking around deployment? Would each OmniSharp plug-in be a separate NuGet package?

The code changes thus far look reasonable. Where could I help out?

**Comment by david-driscoll** [comment]
So I hadn't yet got to deployment, but in reality it would be nice if we could ship them as NuGet Packages.

What I have so far in the PR is:

* `OmniSharp.MsBuild`
  * Depends on
    * `OmniSharp.Abstractions`
    * `OmniSharp.Roslyn.CSharp`
* `OmniSharp.Dnx`
  * Depends on
    * `OmniSharp.Abstractions`
    * `OmniSharp.Roslyn.CSharp`
* `OmniSharp.ScriptCs`
  * Depends on
    * `OmniSharp.Abstractions`
    * `OmniSharp.Roslyn.CSharp`

Right now `OmniSharp` then depends on all of those, so they get loaded up by Dnx.  Today we can pick and choose which ones to load into MEF, but I would prefer something where we could dynamically load the assemblies, based on what we send to the server.  The editor, might know that it's dealing with a dnx or msbuild solution, so we only really need to load in the specific solution.  There could also be mixed solutions, with C#, ScriptCs and F# at the same time (theoretically) so we'd need to be able to load more than one plugin system at a time as well.

What might be interesting is if we could somehow load in the NuGet package inside of `Program` before we bootstrap `Startup`.

This is just my thoughts on what we could possibly do.  Basically add in command line options for where we can get plugins from.  They could be a directory off disc or it could be a NuGet feed or we could default to nuget.org.  If using nuget, restore those packages into the OmniSharp install location itself.  If it's a directory on disc, hopefully we could just leave them in place.  Now the problem is once we restore them, I have _no_ idea how to get (or if it's possible) Dnx to side load those additional nuget packages, since they won't be part of the original project[.lock].json.  Or is there something we can add to the `HostingContext` that would let us add additional assemblies?

We _could_ have an OmniSharp bootstrapper, that does the nuget magic and then writes a project.json out.  That still seems like a lot of work, and would add an extra process to proxy through.

**Comment by david-driscoll** [comment]
Also thoughts, on breaking out the plugin distribution / loading into another issue?

That way we get what we have today, working well, get that merged in.  Then deal with how do we best load in additional user plugins.   Ideally with this merged in, existing editors would not have to make any changes to consume the refactored server.

**Comment by nosami** [comment]
Yeah... plugin distribution should be a separate issue.

File: src/OmniSharp.Abstractions/project.json [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    "Microsoft.Framework.Runtime.Interfaces": "1.0.0-beta4",
+    "Microsoft.Framework.Logging.Console": "1.0.0-beta4",
+    "Microsoft.Framework.Caching.Memory": "1.0.0-beta4",
+    "Microsoft.Framework.FileSystemGlobbing": "1.0.0-beta4",
+    "Microsoft.Composition": "1.0.30"
+  },
+  "frameworks": {
+    "dnx451": {
+      "frameworkAssemblies": {
+        "System.Runtime": "4.0.0.0",
```
Comment (davidfowl) [inline src/OmniSharp.Abstractions/project.json]: Get rid of these versions

File: src/OmniSharp.Abstractions/project.json [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+        "System.Linq": "4.0.0.0",
+        "System.Text.Encoding":"4.0.0.0",
+        "System.IO":"4.0.0.0",
+        "System.Reflection":"4.0.0.0",
+        "System.Security":"4.0.0.0"
+      }
+    },
+    "dnxcore50": {
+      "dependencies": {
+        "System.Runtime": "4.0.20-beta-22816",
```
Comment (davidfowl) [inline src/OmniSharp.Abstractions/project.json]: We should make these versions consistent

File: src/OmniSharp.Plugins/OopResponse.cs [hunk 13]
```diff
@@ -0,0 +1,47 @@
+using System;
+using Newtonsoft.Json.Linq;
+
+namespace OmniSharp.Plugins
+{
+    class OopResponse
```
Comment (davidfowl) [inline src/OmniSharp.Plugins/OopResponse.cs]: Change the name of this to plugin response

File: src/OmniSharp.Plugins/OutOfProcessPlugin.cs [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using OmniSharp.Models.v1;
+using OmniSharp.Services;
+using OmniSharp.Stdio.Protocol;
+using OmniSharp.Stdio.Services;
+
+namespace OmniSharp.Plugins
+{
+    public class OutOfProcessPlugin : IProjectSystem, IDisposable
```
Comment (davidfowl) [inline src/OmniSharp.Plugins/OutOfProcessPlugin.cs]: All plugins are out of process right? Call it a plugin?

File: src/OmniSharp.Plugins/OutOfProcessPlugin.cs [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+        {
+            while (!_cancellation.IsCancellationRequested)
+            {
+                var line = await _process.StandardOutput.ReadLineAsync();
+                if (line == null)
+                {
+                    break;
+                }
+
+                var ignored = Task.Factory.StartNew((Action)(() =>
```
Comment (davidfowl) [inline src/OmniSharp.Plugins/OutOfProcessPlugin.cs]: Why are you casting to an action.

File: src/OmniSharp.Plugins/OutOfProcessPlugin.cs [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+                            _writer.WriteLine(new EventPacket()
+                            {
+                                Event = "error",
+                                Body = response.Message,
+                            });
+                            return;
+                        }
+
+                        Action<string> requestHandler = null;
+                        if (!_requests.TryGetValue((int)response.Request_seq, out requestHandler))
```
Comment (davidfowl) [inline src/OmniSharp.Plugins/OutOfProcessPlugin.cs]: Why are you casting an int to and int?

File: src/OmniSharp.Plugins/OutOfProcessPlugin.cs [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+                        {
+                            throw new ArgumentException("invalid seq-value");
+                        }
+
+                        requestHandler((string)response.BodyJson);
```
Comment (davidfowl) [inline src/OmniSharp.Plugins/OutOfProcessPlugin.cs]: Same question here. Isn't BodyJson already a string?

File: src/OmniSharp.Plugins/OutOfProcessPlugin.cs [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+        }
+
+        public void Dispose()
+        {
+            _cancellation.Cancel();
+        }
+
+        public void Initalize(IConfiguration configuration)
+        {
+            Task.Run(() => Run());
```
Comment (davidfowl) [inline src/OmniSharp.Plugins/OutOfProcessPlugin.cs]: Fire an Forget for a specific reason?

File: src/OmniSharp.Roslyn.CSharp/Services/FindSymbolsService.cs [hunk 19]
```diff
@@ -1,18 +1,27 @@
...
+    [OmniSharpEndpoint(typeof(RequestHandler<FindSymbolsRequest, QuickFixResponse>), LanguageNames.CSharp)]
+    public class FindSymbolsService : RequestHandler<FindSymbolsRequest, QuickFixResponse>
     {
-        [HttpPost("findsymbols")]
-        public async Task<QuickFixResponse> FindSymbols(FindSymbolsRequest request = null)
+        private OmnisharpWorkspace _workspace;
+
+        [ImportingConstructor]
+        public FindSymbolsService(OmnisharpWorkspace workspace) {
+_workspace = workspace;
```
Comment (davidfowl) [inline src/OmniSharp.Roslyn.CSharp/Services/FindSymbolsService.cs]: Formatting?

File: src/OmniSharp.Roslyn.CSharp/Services/GotoDefinitionService.cs [hunk 20]
```diff
@@ -1,20 +1,30 @@
...
-        public async Task<GotoDefinitionResponse> GotoDefinition(GotoDefinitionRequest request)
+        private OmnisharpWorkspace _workspace;
+
+        [ImportingConstructor]
+        public GotoDefinitionService(OmnisharpWorkspace workspace)
+        {
+            _workspace = workspace;
+        }
+
+        async Task<GotoDefinitionResponse> RequestHandler<GotoDefinitionRequest, GotoDefinitionResponse>.Handle(GotoDefinitionRequest request)
```
Comment (davidfowl) [inline src/OmniSharp.Roslyn.CSharp/Services/GotoDefinitionService.cs]: private? Public? internal?

File: src/OmniSharp/Api/v1/Navigation/OmnisharpController.FindUsages.cs [hunk 21]
```diff
@@ -15,41 +15,48 @@ public partial class OmnisharpController
         [HttpPost("findusages")]
         public async Task<QuickFixResponse> FindUsages(FindUsagesRequest request)
         {
-            var document = _workspace.GetDocument(request.FileName);
```
Comment (davidfowl) [inline src/OmniSharp/Api/v1/Navigation/OmnisharpController.FindUsages.cs]: Is this an unrelated change to the plugin model changes?

File: src/OmniSharp/Api/v1/ProjectSystem/OmnisharpController.ProjectSystem.cs [hunk 22]
```diff
@@ -1,66 +1,83 @@
...
-        public WorkspaceInformationResponse ProjectInformation(ProjectInformationRequest request)
+        public async Task<WorkspaceInformationResponse> ProjectInformation(ProjectInformationRequest request)
         {
-            return new WorkspaceInformationResponse
+            var tasks = new List<Task<Action>>();
+            var response = new WorkspaceInformationResponse();
+
+            foreach (var projectSystem in _projectSystems)
+            {
+                tasks.Add(projectSystem.GetInformationModel(request)
```
Comment (davidfowl) [inline src/OmniSharp/Api/v1/ProjectSystem/OmnisharpController.ProjectSystem.cs]: Why are you using continue with?

File: src/OmniSharp/Api/v1/ProjectSystem/OmnisharpController.ProjectSystem.cs [hunk 23]
```diff
@@ -1,66 +1,83 @@
...
-            DnxProject dnxProjectItem = null;
+            var tasks = new List<Task<Action>>();
+            var response = new ProjectInformationResponse();
 
-            if (msBuildContextProject != null)
+            foreach (var projectSystem in _projectSystems)
             {
-                msBuildProjectItem = new MSBuildProject(msBuildContextProject);
+                tasks.Add(projectSystem.GetProjectModel(request.FileName)
+                    .ContinueWith((project) =>
```
Comment (davidfowl) [inline src/OmniSharp/Api/v1/ProjectSystem/OmnisharpController.ProjectSystem.cs]: Same question here.

File: src/OmniSharp/Middleware/Endpoint/EndpointHandler.cs [hunk 24]
```diff
@@ -0,0 +1,281 @@
...
+using System.Threading.Tasks;
+using Microsoft.AspNet.Http;
+using Microsoft.Framework.Logging;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using OmniSharp.Mef;
+using OmniSharp.Models;
+using OmniSharp.Plugins;
+
+namespace OmniSharp.Middleware.Endpoint
```
Comment (davidfowl) [inline src/OmniSharp/Middleware/Endpoint/EndpointHandler.cs]: Way too many classes in this file. Break it up.

File: src/OmniSharp/Middleware/EndpointMiddleware.cs [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+using Newtonsoft.Json;
+using OmniSharp.Mef;
+using OmniSharp.Middleware.Endpoint;
+using OmniSharp.Models;
+using OmniSharp.Plugins;
+using OmniSharp.Services;
+
+namespace OmniSharp.Middleware
+{
+    // You may need to install the Microsoft.AspNet.Http.Abstractions package into your project
```
Comment (davidfowl) [inline src/OmniSharp/Middleware/EndpointMiddleware.cs]: Yes this is true :smile:

File: src/OmniSharp/Startup.cs [hunk 26]
```diff
@@ -119,11 +116,70 @@ public static OmnisharpWorkspace CreateWorkspace()
             return new OmnisharpWorkspace(MefHostServices.Create(assemblies));
         }
 
+        public static CompositionHost ConfigurePluginHost(IServiceProvider serviceProvider,
+                              OmnisharpWorkspace workspace,
```
Comment (davidfowl) [inline src/OmniSharp/Startup.cs]: Fix formatting

File: src/OmniSharp.Abstractions/project.json (hunk 12 above)
Comment (david-driscoll) [inline src/OmniSharp.Abstractions/project.json]: done, pinned `22816` for now.

File: src/OmniSharp.Plugins/OutOfProcessPlugin.cs (hunk 14 above)
Comment (david-driscoll) [inline src/OmniSharp.Plugins/OutOfProcessPlugin.cs]: Done on all counts.

File: src/OmniSharp.Roslyn.CSharp/Services/GotoDefinitionService.cs (hunk 20 above)
Comment (david-driscoll) [inline src/OmniSharp.Roslyn.CSharp/Services/GotoDefinitionService.cs]: Was implemented explicitly, moved to public.

File: src/OmniSharp/Api/v1/Navigation/OmnisharpController.FindUsages.cs (hunk 21 above)
Comment (david-driscoll) [inline src/OmniSharp/Api/v1/Navigation/OmnisharpController.FindUsages.cs]: This was an odd duck... I was getting a new exception, it appears to be related to something in Roslyn itself.  It fails on both the current version of the server and the PR, I'll roll this back.

File: src/OmniSharp/Api/v1/ProjectSystem/OmnisharpController.ProjectSystem.cs (hunk 22 above)
Comment (david-driscoll) [inline src/OmniSharp/Api/v1/ProjectSystem/OmnisharpController.ProjectSystem.cs]: I was concerned that there could be a threading issue, adding the items into the dictionary.

File: src/OmniSharp/Middleware/EndpointMiddleware.cs (hunk 25 above)
Comment (david-driscoll) [inline src/OmniSharp/Middleware/EndpointMiddleware.cs]: copy and paste fail :sheep:

**Comment by david-driscoll** [comment]
holycrap! All tests passed on Travis finally! (the tests are flaky on windows, and I have no idea why, but they are currently flaky on appveyor as well)

cc @davidfowl @DustinCampbell @nosami

**Comment by david-driscoll** [comment]
Merged with master to bring things up to date, also fixed an intellisense bug that has been bothering me.

File: src/OmniSharp.Abstractions/Models/v1/FilesChangedRequest.cs [hunk 27]
```diff
@@ -0,0 +1,8 @@
+using System.Collections.Generic;
+using OmniSharp.Mef;
+
+namespace OmniSharp.Models
+{
+    [OmniSharpEndpoint("/filesChanged", typeof(IEnumerable<Request>), typeof(object), TakeOne = true)]
```
Comment (nosami) [inline src/OmniSharp.Abstractions/Models/v1/FilesChangedRequest.cs]: Feels wrong to have this metadata here (and the other requests)

File: src/OmniSharp.Roslyn.CSharp/Services/Buffer/ChangeBufferService.cs [hunk 28]
```diff
@@ -0,0 +1,26 @@
+using System.Composition;
+﻿using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using OmniSharp.Mef;
+using OmniSharp.Models;
+
+namespace OmniSharp.Roslyn.CSharp.Services.Buffer
+{
+    [OmniSharpHandler(typeof(RequestHandler<ChangeBufferRequest, object>), LanguageNames.CSharp)]
```
Comment (nosami) [inline src/OmniSharp.Roslyn.CSharp/Services/Buffer/ChangeBufferService.cs]: Seems wrong to have this twice `typeof(RequestHandler<ChangeBufferRequest, object>)` Can we use the route name here instead? `/changebuffer` etc.

Comment (david-driscoll) [inline src/OmniSharp.Roslyn.CSharp/Services/Buffer/ChangeBufferService.cs]: We could, but then we'd lose some of the runtime analysis that Mef can give us, since it validates that the ContractType (the handler) matches the type that the attribute is attached to.

File: src/OmniSharp.Abstractions/Models/v1/FilesChangedRequest.cs (hunk 27 above)
Comment (david-driscoll) [inline src/OmniSharp.Abstractions/Models/v1/FilesChangedRequest.cs]: So this is worth some discussion.

I did have a static collection before and it worked fairly well, but then I got thinking about how we could have editors extend OmniSharp with extra features that don't necessarily fit as part of OmniSharp proper.

The big example I have in my head, that I'd like to take a stab at in the future, is Continuous Testing. Using the attribute based approach here lets us add arbitrary endpoints for the editor to call through a custom plugin that editor supports or ships with.

**Comment by nosami** [comment]
What does isMergeable mean? https://github.com/OmniSharp/omnisharp-roslyn/blob/plugins/mef/src/OmniSharp/Middleware/Endpoint/EndpointHandler.cs#L163

Is that used for batching requests?

**Comment by nosami** [comment]
EndpointHandler seems too complex to me although I could be overlooking something.

**Comment by nosami** [comment]
What's isSpreadable for? https://github.com/OmniSharp/omnisharp-roslyn/blob/plugins/mef/src/OmniSharp/Middleware/Endpoint/EndpointHandler.cs#L154

File: src/OmniSharp.Roslyn.CSharp/Services/Buffer/ChangeBufferService.cs (hunk 28 above)
Comment (nosami) [inline src/OmniSharp.Roslyn.CSharp/Services/Buffer/ChangeBufferService.cs]: There is no way to do this other than by adding redundant information?

Comment (david-driscoll) [inline src/OmniSharp.Roslyn.CSharp/Services/Buffer/ChangeBufferService.cs]: It _may_ be possible, but it would outside how Mef seems to do things typically.  In theory we could take the string, check it against the list of end points, and then reflect over the type to ensure that it implements the correct interface.
It _may_ be possible, but it would outside how Mef seems to do things typically.  In theory we could take the string, check it against the list of end points, and then reflect over the type to ensure that it implements the correct interface.

**Comment by nosami** [comment]
Everything looks good to me except for EndpointHandler, which looks more complicated than it needs to be. Not sure why need to use reflection at all for this - MEF should handle that

Could you do something like this (pseudo code-ish)? I've mentioned this before but it got lost in the trail of dust :)
```cs

/// Handles every route / request
public class OmniSharpController<TRequest, TResponse>
{
     public OmniSharpController(Dictionary<routeinfo, IHandler> handlers)
    {
    }

    public TResult Handle(whatever routeInfo, TRequest request)
    {
        return _handlers[routeInfo].Handle(request);
    } 
}
```

**Comment by nosami** [comment]
Having said that, if nobody else comments here you should just merge it anyway (assuming it works!)

**Comment by david-driscoll** [comment]
I could break the endpoint handler into a set of static helper classes for some of the things.

* `_isSpreadable` probably could use a better name.  Basically there is a subset of requests that are special.
  * First there are mergable requests (below)
  * The other are requests that aren't specific to a language, `/project`, `/projects`, `/packagesearch` are examples of requests that don't really belong to a specific language.  So these get picked up and which even language returns a result first, is the response that is sent back.
`_isMergable` are for requests that can be stitched together, for things like getting all diagnostics, or finding symbols.  These will delegate to all plugins and then `IMergableResponse` allows everything to be merged for sending the final response to the client.

File: src/OmniSharp.Roslyn.CSharp/Services/Buffer/ChangeBufferService.cs (hunk 28 above)
Comment (nosami) [inline src/OmniSharp.Roslyn.CSharp/Services/Buffer/ChangeBufferService.cs]: This would look worse in F# where you don't expect to see so much type cluttering.

**Comment by nosami** [comment]
> The other are requests that aren't specific to a language, /project, /projects, /packagesearch are examples of requests that don't really belong to a specific language. So these get picked up and which even language returns a result first, is the response that is sent back.

This doesn't sound right to me. I think you need to know at request time enough information to know which handler will run.

File: src/OmniSharp.Roslyn.CSharp/Services/Buffer/ChangeBufferService.cs (hunk 28 above)
Comment (david-driscoll) [inline src/OmniSharp.Roslyn.CSharp/Services/Buffer/ChangeBufferService.cs]: F# can also do this form, not sure if it's any easier to do in F# or not.

````
        [OmniSharpHandler(typeof(Func<UpdateBufferRequest, Task<object>>), LanguageNames.CSharp)]
        public Func<UpdateBufferRequest, Task<object>> UpdateBuffer { get; } = (request) => { return Task.FromResult<object>(true); };
```

**Comment by david-driscoll** [comment]
I can see if I can factor them out, there have been some changes since it went in.

File: src/OmniSharp.Roslyn.CSharp/Services/Buffer/ChangeBufferService.cs (hunk 28 above)
Comment (nosami) [inline src/OmniSharp.Roslyn.CSharp/Services/Buffer/ChangeBufferService.cs]: I'm not saying it can't be done, just that it would look weird.

In my head, every handler should implement IHandler and have a Handle method.

**Comment by david-driscoll** [comment]
made a small change, and then took some time this weekend to get linux working in my VM, and using this branch everything seems to be working.

image

**Comment by nosami** [comment]
Confirmed working with vim over HTTP

**Comment by nosami** [comment]
Epic! :fireworks: :dancer: :dancers: :dash:
